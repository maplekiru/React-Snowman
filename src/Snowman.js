import React, { useState } from "react";

import "./Snowman.css";
import {ENGLISH_WORDS , randomWord } from "./words";
import img0 from "./0.png";
import img1 from "./1.png";
import img2 from "./2.png";
import img3 from "./3.png";
import img4 from "./4.png";
import img5 from "./5.png";
import img6 from "./6.png";


/** Snowman game: plays hangman-style game with a melting snowman.
 *
 * Props:
 * - maxWrong: how many wrong moves is a player allowed?
 * - images: array of images for wrong guess
 * - words: array of words to pick answer from
 *
 * State:
 * - nWrong: # wrong guesses so far
 * - guessed: set of guessed letters (good and bad) so far
 * - answer: selected secret word*
 */

function Snowman(props) {
  /** by default, allow 6 guesses and use provided gallows images. */

  const [nWrong, setNWrong] = useState(0);
  const [guessed, setGuessed] = useState(new Set());
  const [answer, setAnswer] = useState(randomWord(props.words));

  /** guessedWord: show current-state of word:
   if guessed letters are {a,p,e}, show "app_e" for "apple"
   */
  function guessedWord() {
    return answer
        .split("")
        .map(ltr => (guessed.has(ltr) ? ltr : "_"));
  }

  /** handleGuess: handle a guessed letter:
   - add to guessed letters
   - if not in answer, increase number-wrong guesses
   */
  function handleGuess(evt) {
    let ltr = evt.target.value;

    setGuessed(g => {
      const newGuessed = new Set(g);
      newGuessed.add(ltr);
      return newGuessed;
    });

    setNWrong(n => n + (answer.includes(ltr) ? 0 : 1));
  }

  /** checks if max wrong has been reached for current game */
  let isMaxWrong = (nWrong === props.maxWrong)

  /** generateButtons: return array of letter buttons to render */
  function generateButtons() {
    return "abcdefghijklmnopqrstuvwxyz".split("").map(ltr => (
        <button
            key={ltr}
            id={ltr}
            value={ltr}
            onClick={handleGuess}
            disabled={guessed.has(ltr)}
        >
          {ltr}
        </button>
    ));
  }

  /** render: render game */
  return (
      <div className="Snowman">
        <img src={(props.images)[nWrong]} alt={nWrong} />
        <p> Number wrong: {nWrong}</p>
        <p className="Snowman-word">{guessedWord()}</p>
        {/* Checks if the max number of wrong guesses reached,
          if it has, then shows "you lose", otherwise, 
          renders list of letters as buttons */}
        { isMaxWrong ? (
          <p>you lose</p>
        ) : (
          <p className="Snowman-buttons">{generateButtons()}</p>
        )}
      </div>
  );
}

Snowman.defaultProps = {
  maxWrong: 6,
  images: [img0, img1, img2, img3, img4, img5, img6],
  words: ENGLISH_WORDS,
};


export default Snowman;