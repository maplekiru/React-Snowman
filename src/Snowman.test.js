import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Snowman from "./Snowman";


it("renders Snowman without crashing", function () {
  render(<Snowman />);
});

it("matched snapshot of Snowman", function () {
  const { container } = render(<Snowman />);
  expect(container).toMatchSnapshot();
});

it("After max wrong # of guesses do not show button and show 'you lose'",
  function () {
    const { container } = render(<Snowman />);
    
    const letterB = container.querySelector("#b");
    fireEvent.click(letterB);
    const letterC = container.querySelector("#c");
    fireEvent.click(letterC);
    const letterZ = container.querySelector("#z");
    fireEvent.click(letterZ);
    const letterD = container.querySelector("#d");
    fireEvent.click(letterD);
    const letterF = container.querySelector("#f");
    fireEvent.click(letterF);
    const letterG = container.querySelector("#g");
    fireEvent.click(letterG);

    expect(
      container.querySelector('.Snowman-buttons')
    ).not.toBeInTheDocument();
    expect(container).toContainHTML("you lose");


  });


