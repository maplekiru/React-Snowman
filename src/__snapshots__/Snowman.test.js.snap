// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`After max wrong # of guesses do not show button and show 'you lose' 1`] = `
<div>
  <div
    class="Snowman"
  >
    <img
      alt="6"
      src="6.png"
    />
    <p>
       Number wrong: 
      6
    </p>
    <p
      class="Snowman-word"
    >
      _
      _
      _
      _
      _
    </p>
    <p>
      you lose
    </p>
  </div>
</div>
`;

exports[`matched snapshot of Snowman 1`] = `
<div>
  <div
    class="Snowman"
  >
    <img
      alt="0"
      src="0.png"
    />
    <p>
       Number wrong: 
      0
    </p>
    <p
      class="Snowman-word"
    >
      _
      _
      _
      _
      _
    </p>
    <p
      class="Snowman-buttons"
    >
      <button
        id="a"
        value="a"
      >
        a
      </button>
      <button
        id="b"
        value="b"
      >
        b
      </button>
      <button
        id="c"
        value="c"
      >
        c
      </button>
      <button
        id="d"
        value="d"
      >
        d
      </button>
      <button
        id="e"
        value="e"
      >
        e
      </button>
      <button
        id="f"
        value="f"
      >
        f
      </button>
      <button
        id="g"
        value="g"
      >
        g
      </button>
      <button
        id="h"
        value="h"
      >
        h
      </button>
      <button
        id="i"
        value="i"
      >
        i
      </button>
      <button
        id="j"
        value="j"
      >
        j
      </button>
      <button
        id="k"
        value="k"
      >
        k
      </button>
      <button
        id="l"
        value="l"
      >
        l
      </button>
      <button
        id="m"
        value="m"
      >
        m
      </button>
      <button
        id="n"
        value="n"
      >
        n
      </button>
      <button
        id="o"
        value="o"
      >
        o
      </button>
      <button
        id="p"
        value="p"
      >
        p
      </button>
      <button
        id="q"
        value="q"
      >
        q
      </button>
      <button
        id="r"
        value="r"
      >
        r
      </button>
      <button
        id="s"
        value="s"
      >
        s
      </button>
      <button
        id="t"
        value="t"
      >
        t
      </button>
      <button
        id="u"
        value="u"
      >
        u
      </button>
      <button
        id="v"
        value="v"
      >
        v
      </button>
      <button
        id="w"
        value="w"
      >
        w
      </button>
      <button
        id="x"
        value="x"
      >
        x
      </button>
      <button
        id="y"
        value="y"
      >
        y
      </button>
      <button
        id="z"
        value="z"
      >
        z
      </button>
    </p>
  </div>
</div>
`;
